device/iot-device-samples/device-reconnection-sample/src/main/java/samples/com/microsoft/azure/sdk/iot/IotHubConnectionStatusChangeCallbackHandler.java

```diff
@@ -0,0 +1,95 @@
+// Copyright (c) Microsoft. All rights reserved.Licensed under the MIT license.
+// See LICENSE file in the project root for full license information.
+
+package samples.com.microsoft.azure.sdk.iot;
+
+import com.microsoft.azure.sdk.iot.device.IotHubConnectionStatusChangeCallback;
+import com.microsoft.azure.sdk.iot.device.IotHubConnectionStatusChangeReason;
+import com.microsoft.azure.sdk.iot.device.transport.IotHubConnectionStatus;
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
+public class IotHubConnectionStatusChangeCallbackHandler implements IotHubConnectionStatusChangeCallback {
+    @Override
+    public void execute(IotHubConnectionStatus status, IotHubConnectionStatusChangeReason statusChangeReason, Throwable throwable, Object callbackContext) {
+        log.debug("### Connection status change reported: status={}, reason={}", status, statusChangeReason);
+
+        if (throwable != null) {
+            log.warn("Connection status callback executed with a throwable");
+            log.warn("Throwable: {}", throwable);
+        }
+
+        final DeviceClientManager clientManager = (DeviceClientManager) callbackContext;
+
+        switch (status) {
```

Comment:
Please don't use `switch`. The code could be much simpler if just handle what is required which is logging status, reason and exception,  reconnect when `DISCONNECTED` + `RETRY_EXPIRED`/`COMMUNICATION_ERROR`

Reply:
+1 on this point. The switch statement here is massive. An If block would be much easier to read

---
In reply to: [340885696](https://github.com/Azure/azure-iot-sdk-java/pull/612#discussion_r340885696) [](ancestors = 340885696)

Reply:
The reason for listing all states and reasons explicitly is to provide the customer with an example which has all of the combinations (even those which the sdk would never reach), and the expected behavior in each case. I chose the `switch` statement for better readability; if you guys feel an `if` statement would be more readable, I can change it to that.